src/ocean/meta/types/ReduceType.d

```diff
@@ -0,0 +1,199 @@
+                return 1;
+            else
+                return 0;
+        }
+    }
+}
+
+unittest
+{
+    static assert (ReduceType!(int, ExampleReducer) == 1);
+    static assert (ReduceType!(int[int], ExampleReducer) == 2);
+}
+
+/*******************************************************************************
+
+    Verifies that `Reducer` is an aggregate type that conforms type reducer
+    requirements and issues static assertion otherwise.
+
+    Params:
+        Reducer = aggregate type to check
+
+*******************************************************************************/
+
+public template assertValidReducer ( Reducer )
+{
+    static assert (
+        is(Reducer.Result),
+        Reducer.stringof ~  " must define `Result` type alias"
+    );
+    static assert (
+        is(typeof(Reducer.seed) == Reducer.Result),
+        Reducer.stringof ~ " must define `seed` constant of type `Result`"
+    );
+    static assert (
+        is(typeof(Reducer.accumulate)),
+        Reducer.stringof ~ " must define `accumluate` CTFE function"
+    );
+    static assert (
+        is(typeof(Reducer.visit!(int))),
+        Reducer.stringof ~ " must define `visit` templated function"
+    );
+
+    alias void assertValidReducer;
+}
+
+/*******************************************************************************
+
+    Implementation for `ReduceType` handling type recursion
+
+    Params:
+        Reducer = see `ReduceType` param documentation
+
+*******************************************************************************/
+
+private struct ReduceTypeImpl ( Reducer )
+{
+    // Give better error messages for wrong `Reducer` definitions
+    alias assertValidReducer!(Reducer) check;
+
+    // Helper to calculate new value and update accumulator in one step
+    static private void accumulate ( T ) ( ref T accum, T next )
+    {
+        accum = Reducer.accumulate(accum, next);
+    }
+
+    // Main recursive visiting implementation
+    static private Reducer.Result reduce ( T ) ( )
+    {
+        auto result = Reducer.seed;
+        accumulate(result, Reducer.visit!(T)());
+
+        static if (isPrimitiveType!(T))
+        {
+            // do nothing, already processed
+        }
+        else static if (isAggregateType!(T))
+        {
+            foreach (TElem; typeof(T.init.tupleof))
+                accumulate(result, reduce!(TElem)());
+        }
+        else static if (isArrayType!(T))
+        {
+            static if (is(T U : U[]))
+            {
+                // dynamic or static
+                accumulate(result, reduce!(U)());
+            }
+            else
+            {
+                // associative
+                accumulate(result, reduce!(typeof(T.init.keys[0]))());
+                accumulate(result, reduce!(typeof(T.init.values[0]))());
+            }
+        }
+        else static if (isPointerType!(T))
+        {
+            accumulate(result, reduce!(typeof(*T.init))());
```

Comment:
- Suggest `is(Unqual!(T) U == U*)` since you need the pointer base type `U` for `reduce`.
- `isPointerType!(T)` yields `true` if `T` is a function pointer so the ` static if (isFunctionType!(T))` should be prior to this branch.